libs/sdk-common/src/input_parser.rs

```diff
@@ -197,6 +225,80 @@ pub async fn parse(
     Err(anyhow!("Unrecognized input type"))
 }
 
+fn get_by_key(tuple_vector: &[(&str, &str)], key: &str) -> Option<String> {
+    tuple_vector
+        .iter()
+        .find(|(k, _)| *k == key)
+        .map(|(_, v)| v.to_string())
+}
+
+fn parse_bip353_record(bip353_record: String) -> Option<String> {
+    let (_, query_part) = bip353_record.split_once("?")?;
+
+    let query_params = querystring::querify(query_part);
+
+    return get_by_key(&query_params, BOLT12_PREFIX)
+        .or_else(|| get_by_key(&query_params, LNURL_PAY_PREFIX));
+}
+
+fn validate_bip353_record(decoded: &str) -> bool {
+    if !decoded.to_lowercase().starts_with(BIP353_PREFIX) {
+        error!(
+            "Invalid decoded TXT data (doesn't begin with: {})",
+            BIP353_PREFIX
+        );
+
+        return false;
+    }
+
+    return true;
+}
+
+fn extract_bip353_record(records: Vec<String>) -> Option<String> {
+    let bip353_record = records
+        .iter()
+        .filter(|record| validate_bip353_record(record))
+        .collect::<Vec<&String>>();
+
+    let bip353_prefix_counter = bip353_record.len();
+    if bip353_prefix_counter != BIP353_PREFIX_COUNT_CONSTRAINT {
+        error!(
+            "Invalid decoded TXT data. Number of {} != {}",
+            BIP353_PREFIX, BIP353_PREFIX_COUNT_CONSTRAINT
+        );
+
+        return None;
+    }
+
+    return bip353_record.first().map(|record| record.to_string());
+}
```

Comment:
Consider the following alternative, which IMO is simpler

- Drops intermediate value `bip353_prefix_counter`
- Drops prefix count constant. It isn't foreseeable that this value changes and it isn't referenced in several places, so I don't think it adds value
- Uses into_iter to avoid the map at the end

```suggestion
fn extract_bip353_record(records: Vec<String>) -> Option<String> {
    let bip353_record = records
        .into_iter()
        .filter(|record| validate_bip353_record(record))
        .collect::<Vec<String>>();

    if bip353_record.len() > 1 {
        error!(
            "Invalid decoded TXT data. Multiple records found ({})",
            bip353_record.len()
        );
        return None;
    }

    bip353_record.first().cloned()
}
```